internal/mturk/step.go

```diff
@@ -189,19 +222,135 @@ func (s *Session) runMTurkMessageStep(ctx context.Context, project *ent.Project,
+			})
+		}).
+		WithRun(func(run *ent.RunQuery) {
+			run.WithSteps()
+		}).
+		Where(stepRunModel.IDEQ(stepRun.ID)).
+		First(ctx)
+	if err != nil {
+		log.Error().Err(err).Msg("get stepRun")
+		return errors.Wrap(err, "get stepRun")
+	}
+
+	step, err = stepRun.Edges.StepOrErr()
+	if err != nil {
+		log.Error().Err(err).Msg("get step")
+		return errors.Wrap(err, "get step")
+	}
+
+	run, err = stepRun.Edges.RunOrErr()
+	if err != nil {
+		log.Error().Err(err).Msg("get run")
+		return errors.Wrap(err, "get run")
 	}
 
-	output, err := s.notifyWorkers(ctx, subject, msg, workerIDs)
+	stepRuns, err := run.Edges.StepsOrErr()
 	if err != nil {
-		return errors.Wrap(err, "notify workers")
+		log.Error().Err(err).Msg("get step runs")
+		return errors.Wrap(err, "get step runs")
 	}
 
+	template, err = step.Edges.TemplateOrErr()
+	if err != nil {
+		log.Error().Err(err).Msg("get template")
+		return errors.Wrap(err, "get template")
+	}
+
+	steps, err := template.Edges.StepsOrErr()
+	if err != nil {
+		log.Error().Err(err).Msg("get steps")
+		return errors.Wrap(err, "get steps")
+	}
+
+	rsteps := make([]*renderStep, len(steps))
+	t := *run.StartedAt
+	for i, s := range steps {
+		var startsAt, startedAt, endedAt string
+		if stepRuns[i].Index <= stepRun.Index {
+			startedAt = stepRun.StartedAt.Format(time.Kitchen)
+			startsAt = startedAt
+			if stepRuns[i].Index < stepRun.Index {
+				endedAt = stepRun.EndedAt.Format(time.Kitchen)
+			}
+		} else {
+			startsAt = t.Format(time.Kitchen)
+			t = t.Add(time.Duration(step.Duration) * time.Minute)
+		}
+		rsteps[i] = &renderStep{
+			Index:             s.Index,
+			Duration:          s.Duration,
+			ParticipantsCount: stepRuns[i].ParticipantsCount,
+			Type:              s.Type.String(),
+			StartsAt:          startsAt,
+			StartedAt:         startedAt,
+			EndedAt:           endedAt,
+		}
+	}
+
+	startedAt := stepRun.StartedAt.Format(time.Kitchen)
 	failedWorkedIDs := make(map[string]*mturk.NotifyWorkersFailureStatus)
-	for _, stat := range output.NotifyWorkersFailureStatuses {
-		failedWorkedIDs[*stat.WorkerId] = stat
+	for _, workerID := range workerIDs {
+		tempWorkerIDs := make([]string, 0, 1)
+		tempWorkerIDs = append(tempWorkerIDs, workerID)
+
+		renderCtx := &renderContext{
+			Template: &renderTemplate{
+				Adult:            template.Adult,
+				Sandbox:          template.Sandbox,
+				SelectionType:    template.SelectionType.String(),
+				Name:             template.Name,
+				ParticipantCount: template.ParticipantCount,
+			},
+			Run: &renderRun{
+				Name:      run.Name,
+				StartedAt: run.StartedAt.Format(time.Kitchen),
+			},
+			Step: &renderStep{
+				Index:             step.Index,
+				Duration:          step.Duration,
+				ParticipantsCount: stepRun.ParticipantsCount,
+				Type:              step.Type.String(),
+				StartsAt:          startedAt,
+				StartedAt:         startedAt,
+			},
+			Steps:    rsteps,
+			URL:      url.String(),
+			WorkerID: workerID,
+		}
```

Comment:
The only things that change in the renderCtx is the workerId and potentially in the future the URL (when we do the programatic URL change), so I don't think we need to rebuild this renderContext object (and most of it's fields) over and over, just once, then replace workerID (and url in the future).